File: rust/kagent/src/soul/approval.rs
use std::collections::{HashMap, HashSet};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, Mutex};

use thiserror::Error;
use tracing::debug;
use uuid::Uuid;

use crate::utils::{Queue, QueueShutDown};
use crate::wire::{ApprovalRequest, ApprovalResponseKind, DisplayBlock};

use super::toolset::get_current_tool_call_or_none;

#[derive(Debug, Error)]
pub enum ApprovalError {
    #[error("approval must be requested from a tool call")]
    NoToolCall,
    #[error("approval queue is shut down")]
    QueueShutDown,
    #[error("no pending approval request with id {0}")]
    RequestNotFound(String),
    #[error("approval response channel closed")]
    ResponseClosed,
}

struct ApprovalState {
    yolo: AtomicBool,
    auto_approve_actions: Mutex<HashSet<String>>,
}

impl ApprovalState {
    fn new(yolo: bool) -> Self {
        Self {
            yolo: AtomicBool::new(yolo),
            auto_approve_actions: Mutex::new(HashSet::new()),
        }
    }
}

pub struct Approval {
    request_queue: Queue<ApprovalRequest>,
    requests: Mutex<HashMap<String, ApprovalRequest>>,
    state: Arc<ApprovalState>,
}

impl Approval {
    pub fn new(yolo: bool) -> Self {
        Self {
            request_queue: Queue::new(),
            requests: Mutex::new(HashMap::new()),
            state: Arc::new(ApprovalState::new(yolo)),
        }
    }

    pub fn set_yolo(&self, yolo: bool) {
        self.state.yolo.store(yolo, Ordering::SeqCst);
    }

    pub fn is_yolo(&self) -> bool {
        self.state.yolo.load(Ordering::SeqCst)
    }

    pub fn share(&self) -> Self {
        Self {
            request_queue: Queue::new(),
            requests: Mutex::new(HashMap::new()),
            state: Arc::clone(&self.state),
        }
    }

    pub async fn request(
        &self,
        sender: &str,
        action: &str,
        description: &str,
        display: Option<Vec<DisplayBlock>>,
    ) -> Result<bool, ApprovalError> {
        let tool_call = get_current_tool_call_or_none().ok_or(ApprovalError::NoToolCall)?;
        debug!(
            "{} ({}) requesting approval: {} {}",
            tool_call.function.name, tool_call.id, action, description
        );

        if self.is_yolo() {
            return Ok(true);
        }

        if self
            .state
            .auto_approve_actions
            .lock()
            .unwrap()
            .contains(action)
        {
            return Ok(true);
        }

        let request = ApprovalRequest::new(
            Uuid::new_v4().to_string(),
            tool_call.id,
            sender,
            action,
            description,
            display.unwrap_or_default(),
        );

        self.request_queue
            .put_nowait(request.clone())
            .map_err(|QueueShutDown| ApprovalError::QueueShutDown)?;

        self.requests
            .lock()
            .unwrap()
            .insert(request.id.clone(), request.clone());

        let response = request.wait().await;
        let approved = match response {
            ApprovalResponseKind::Approve => true,
            ApprovalResponseKind::ApproveForSession => {
                self.state
                    .auto_approve_actions
                    .lock()
                    .unwrap()
                    .insert(request.action.clone());
                true
            }
            ApprovalResponseKind::Reject => false,
        };
        Ok(approved)
    }

    pub async fn fetch_request(&self) -> Result<ApprovalRequest, ApprovalError> {
        loop {
            let request = self
                .request_queue
                .get()
                .await
                .map_err(|QueueShutDown| ApprovalError::QueueShutDown)?;

            if self
                .state
                .auto_approve_actions
                .lock()
                .unwrap()
                .contains(&request.action)
            {
                debug!(
                    "Auto-approving previously requested action: {}",
                    request.action
                );
                let _ = self.resolve_request(&request.id, ApprovalResponseKind::Approve);
                continue;
            }

            return Ok(request);
        }
    }

    pub fn resolve_request(
        &self,
        request_id: &str,
        response: ApprovalResponseKind,
    ) -> Result<(), ApprovalError> {
        let request = {
            let mut requests = self.requests.lock().unwrap();
            requests
                .remove(request_id)
                .ok_or_else(|| ApprovalError::RequestNotFound(request_id.to_string()))?
        };

        if matches!(response, ApprovalResponseKind::ApproveForSession) {
            self.state
                .auto_approve_actions
                .lock()
                .unwrap()
                .insert(request.action.clone());
        }

        debug!(
            "Received approval response for request {}: {:?}",
            request_id, response
        );
        request.resolve(response);
        Ok(())
    }
}

File: src/kimi_cli/soul/approval.py
from __future__ import annotations

import uuid
from collections.abc import Callable
from typing import Literal

from kimi_cli.approval_runtime import (
    ApprovalCancelledError,
    ApprovalRuntime,
    ApprovalSource,
    get_current_approval_source_or_none,
)
from kimi_cli.soul.toolset import get_current_tool_call_or_none
from kimi_cli.tools.utils import ToolRejectedError
from kimi_cli.utils.logging import logger
from kimi_cli.wire.types import DisplayBlock

type Response = Literal["approve", "approve_for_session", "reject"]


class ApprovalResult:
    """Result of an approval request. Behaves as bool for backward compatibility."""

    __slots__ = ("approved", "feedback")

    def __init__(self, approved: bool, feedback: str = ""):
        self.approved = approved
        self.feedback = feedback

    def __bool__(self) -> bool:
        return self.approved

    def rejection_error(self) -> ToolRejectedError:
        if self.feedback:
            return ToolRejectedError(
                message=(f"The tool call is rejected by the user. User feedback: {self.feedback}"),
                brief=f"Rejected: {self.feedback}",
                has_feedback=True,
            )
        source = get_current_approval_source_or_none()
        is_subagent = source is not None and source.agent_id is not None
        if is_subagent:
            return ToolRejectedError(
                message=(
                    "The tool call is rejected by the user. "
                    "Try a different approach to complete your task, or explain the "
                    "limitation in your summary if no alternative is available. "
                    "Do not retry the same tool call, and do not attempt to bypass "
                    "this restriction through indirect means."
                ),
            )
        return ToolRejectedError()


class ApprovalState:
    def __init__(
        self,
        yolo: bool = False,
        auto_approve_actions: set[str] | None = None,
        on_change: Callable[[], None] | None = None,
    ):
        self.yolo = yolo
        self.auto_approve_actions: set[str] = auto_approve_actions or set()
        """Set of action names that should automatically be approved."""
        self._on_change = on_change

    def notify_change(self) -> None:
        if self._on_change is not None:
            self._on_change()


class Approval:
    def __init__(
        self,
        yolo: bool = False,
        *,
        state: ApprovalState | None = None,
        runtime: ApprovalRuntime | None = None,
    ):
        self._state = state or ApprovalState(yolo=yolo)
        self._runtime = runtime or ApprovalRuntime()

    def share(self) -> Approval:
        """Create a new approval queue that shares state (yolo + auto-approve)."""
        return Approval(state=self._state, runtime=self._runtime)

    def set_runtime(self, runtime: ApprovalRuntime) -> None:
        self._runtime = runtime

    @property
    def runtime(self) -> ApprovalRuntime:
        return self._runtime

    def set_yolo(self, yolo: bool) -> None:
        self._state.yolo = yolo
        self._state.notify_change()

    def is_yolo(self) -> bool:
        return self._state.yolo

    async def request(
        self,
        sender: str,
        action: str,
        description: str,
        display: list[DisplayBlock] | None = None,
    ) -> ApprovalResult:
        """
        Request approval for the given action. Intended to be called by tools.

        Args:
            sender (str): The name of the sender.
            action (str): The action to request approval for.
                This is used to identify the action for auto-approval.
            description (str): The description of the action. This is used to display to the user.

        Returns:
            ApprovalResult: Result with ``approved`` flag and optional ``feedback``.
                Behaves as ``bool`` via ``__bool__``, so ``if not result:`` works.

        Raises:
            RuntimeError: If the approval is requested from outside a tool call.
        """
        tool_call = get_current_tool_call_or_none()
        if tool_call is None:
            raise RuntimeError("Approval must be requested from a tool call.")

        logger.debug(
            "{tool_name} ({tool_call_id}) requesting approval: {action} {description}",
            tool_name=tool_call.function.name,
            tool_call_id=tool_call.id,
            action=action,
            description=description,
        )
        if self._state.yolo:
            from kimi_cli.telemetry import track

            track(
                "tool_approved",
                tool_name=tool_call.function.name,
                approval_mode="yolo",
            )
            return ApprovalResult(approved=True)

        if action in self._state.auto_approve_actions:
            from kimi_cli.telemetry import track

            track(
                "tool_approved",
                tool_name=tool_call.function.name,
                approval_mode="auto_session",
            )
            return ApprovalResult(approved=True)

        request_id = str(uuid.uuid4())
        display_blocks = display or []
        source = get_current_approval_source_or_none() or ApprovalSource(
            kind="foreground_turn",
            id=tool_call.id,
        )
        self._runtime.create_request(
            request_id=request_id,
            tool_call_id=tool_call.id,
            sender=sender,
            action=action,
            description=description,
            display=display_blocks,
            source=source,
        )
        try:
            response, feedback = await self._runtime.wait_for_response(request_id)
        except ApprovalCancelledError:
            from kimi_cli.telemetry import track

            track(
                "tool_rejected",
                tool_name=tool_call.function.name,
                approval_mode="cancelled",
            )
            record = self._runtime.get_request(request_id)
            return ApprovalResult(approved=False, feedback=record.feedback if record else "")
        from kimi_cli.telemetry import track

        match response:
            case "approve":
                track(
                    "tool_approved",
                    tool_name=tool_call.function.name,
                    approval_mode="manual",
                )
                return ApprovalResult(approved=True)
            case "approve_for_session":
                track(
                    "tool_approved",
                    tool_name=tool_call.function.name,
                    approval_mode="manual",
                )
                self._state.auto_approve_actions.add(action)
                self._state.notify_change()
                for pending in self._runtime.list_pending():
                    if pending.action == action:
                        self._runtime.resolve(pending.id, "approve")
                return ApprovalResult(approved=True)
            case "reject":
                track(
                    "tool_rejected",
                    tool_name=tool_call.function.name,
                    approval_mode="manual",
                )
                return ApprovalResult(approved=False, feedback=feedback)
            case _:
                track(
                    "tool_rejected",
                    tool_name=tool_call.function.name,
                    approval_mode="manual",
                )
                return ApprovalResult(approved=False)
